### DISCUSSION TITLE
looping through the message (response) data

### DISCUSSION THREAD

**Discussion by TLS1000** [eModbus/eModbus#68] [Q&A]
Hi, first of all great work creating this library; the ESP32 with arduino core was in need for this. For years I've been using the 4-20mA modbus library, but for my new application I'm running into trouble with the multi-tasking.

I would like to mirror a few hundred holding registers from a RTU host/slave into my own uint16_T[] arrays. The data is in about 10 series of successive ranges. 

From another discussion, I understood that I can use different tokens in the request, so threre can be different handlers that copy the received data to the designated array.

I have a question about looping through the message data (array of words). I would like to use more basic C++ commands that are a better match for my own C++knowledge; could I use in the handler functions something like this:
```cpp
for(int i=0;i<msg.dataLen/2;i++){  
   myWordarray[i]=msg.get(i,double&);
}
```

thank you

**Comment by Miq1** [comment]
Thanks for the flowers! 🤣

You are almost there. I would suggest you wrote something along
```
uint16_t posInMsg = 3;  // first byte behind the length byte - supposing FC 03 has been used
uint16_t arrayIndex = 0;
double values[24];  // or whatever size you want
for (posInMsg = 3; posInMsg < msg.size(); posInMsg += sizeof(double)) {
  msg.get(posInMsg, values[arrayIndex++]);
}
```
Please note that this is assuming the server having sent the ``double`` values in IEEE754 order, each spanning 4 consecutive 16bit words.

**Reply by TLS1000** [reply]
well understood, thank you

**Reply by TLS1000** [reply]
I'm using the system above, and it works. However when I do the readholdingregister function for more than 29 words, I get a CRC error on about 1 out of four responses (modbusmessage length>61bytes gives problem).    When I only ask for 29 registers, it runs flawless even at high rate. The baudrate my device is using is a low 9600bps. The slave device I'm polling has been tested with previous modbus library (non async)and worked flawless returning 64 words-128 bytes per cycle.

[1;33m[E] 37624| emodbustest.ino      [  45] handleError: Error response: E2 - CRC check error
[0m[1;33m[E] 38124| emodbustest.ino      [  45] handleError: Error response: E2 - CRC check error
[0m[1;33m[E] 38624| emodbustest.ino      [  45] handleError: Error response: E2 - CRC check error
[0m[N] 39210| emodbustest.ino      [  29] handleData: Response: serverID=1, FC=3, Token=000004A6, length=123:
[N] Data: @3FFB9818/123:
  | 0000: 01 03 78 00 0A 00 00 00  00 00 27 00 00 00 00 00  |..x.......'.....|
 | 0010: 77 00 00 09 23 0F D3 00  00 00 80 00 00 09 2D 0F  |w...#.........-.|
| 0020: E5 00 01 00 7E 00 02 09  27 0F DA 00 00 13 8A FC  |....~...'.......|

any idea what this could cause?

**Reply by Miq1** [reply]
Can you give us some more information on your sketch? I assume the ``emodbustest.ino [45]`` is pointing to your ``onError`` callback function?

The Arduino IDE unfortunately does not support preprocessor directives in the compiler call, so you may open the ``Logging.h`` file in your local copy of the eModbus lib and change the line with the default ``LOG_LEVEL`` to ``LOG_LEVEL_VERBOSE``. If you will build your sketch again with this modification, it will print out tons of debug info that might be helpful finding the root cause.

**Reply by Miq1** [reply]
Another observation: the length of the response is reported as ``123``. 29 registers (Modbus words) are 58 bytes only, plus the three header bytes (server ID, FC and length) make 61! Can it be you are mixing Modbus words (16 bit) with ESP32 words (32 bits) somehow?

**Reply by Miq1** [reply]
And a final one: did you notice this **important** remark in the docs: [eModbus Requirements](https://emodbus.github.io/requirements#important-note-for-using-the-rtu-bridge-server-and-client)?

**Reply by TLS1000** [reply]
with verbose mode on:

```
[1;33m[E] 2166| emodbustest.ino      [  45] handleError: Error response: E2 - CRC check error
[0m[V] 2558| ModbusMessage.cpp    [ 505] checkData: Check data #3
[D] 2558| ModbusClientRTU.cpp  [ 106] addToQueue: RC=01
[D] 2559| ModbusClientRTU.cpp  [ 124] handleConnection: Pulled request from queue
[D] 2573| ModbusClientRTU.cpp  [ 129] handleConnection: Request sent.
[D] 2653| ModbusClientRTU.cpp  [ 134] handleConnection: Data response received.
[D] 2653| ModbusClientRTU.cpp  [ 162] handleConnection: Response generated.
[V] Response packet: @3FFB9948/3:
  | 0000: 01 83 E2                                          |...             |

```

**Reply by TLS1000** [reply]
> Another observation: the length of the response is reported as `123`. 29 registers (Modbus words) are 58 bytes only, plus the three header bytes (server ID, FC and length) make 61! Can it be you are mixing Modbus words (16 bit) with ESP32 words (32 bits) somehow?

I don't think that's it: when I'm asking for 60 words the indicated size is 123 so that seems correct

**Reply by TLS1000** [reply]
> And a final one: did you notice this **important** remark in the docs: [eModbus Requirements](https://emodbus.github.io/requirements#important-note-for-using-the-rtu-bridge-server-and-client)?

no I haven't done this. Is it required for modbus RTU?

**Reply by Miq1** [reply]
So we can assume the server is responding with a packet whose CRC does not fit - or ``ModbusClientRTU`` is not getting the complete message. Please see the last remark I made: is your ``esp32-hal-uart.c`` patched as advised?

Crossing posts... Yeah, for best results please patch the core file accordingly.

**Reply by TLS1000** [reply]
> So we can assume the server is responding with a packet whose CRC does not fit - or `ModbusClientRTU` is not getting the complete message. Please see the last remark I made: is your `esp32-hal-uart.c` patched as advised?
> 
> Crossing posts... Yeah, for best results please patch the core file accordingly.

it seems uncommented by default? I downloaded the zip today:

```
// Workaround: uncomment the following line to wait for 16ms(!) for the handling to finish:
      #if RXFIFO_FULL_THRHD_PATCHED == 0
      if (micros() - lastMicros >= 16000) {
      #endif
```

**Reply by Miq1** [reply]
There is a script in the respective subfolder to patch the core automatically, but unfortunately that will not work with the Arduino IDE either. We are normally using PlatformIO only to develop, where the script is applied.

So it is left to your hands to patch the files manually and set the ``RXFIFO_FULL_THRD_PATCHED`` define to 1 in ``options.h``  - sorry for that.

**Reply by TLS1000** [reply]
I set the patch to 1, but now ALL responses return CRC error...

**Reply by Miq1** [reply]
You need to also alter the ``esp32-hal-uart.c`` file - did you do it?

**Reply by TLS1000** [reply]
I'm looking for that file, don't know in which folder it is...

**Reply by Miq1** [reply]
I do not know, either, with the Arduino IDE...

On Windows, it might be something like
```
C:\Users\<user>\Documents\Arduino\hardware\espressif\arduino-esp32-master\cores\esp32\esp32-hal-uart.c
```
I would suggest a file search for the name - it should be found in a few places only.

**Reply by TLS1000** [reply]
I made the modification of the hal-uart file and the option file as instructed but problem got worse: every message has CRC error now.

**Reply by TLS1000** [reply]
my complete sketch:

```
// =================================================================================================
// eModbus: Copyright 2020 by Michael Harwerth, Bert Melis and the contributors to ModbusClient
//               MIT license - see license.md for details
// =================================================================================================

// Example code to show the usage of the eModbus library. 
// Please refer to root/Readme.md for a full description.

// Includes: <Arduino.h> for Serial etc.
#include <Arduino.h>
//#include <HardwareSerial.h>
//HardwareSerial serial1(1);

// Include the header for the ModbusClient RTU style
#include "ModbusClientRTU.h"

// For demonstration, use the LOG statements for output
#include "Logging.h"

// Create a ModbusRTU client instance
// In my case the RS485 module had auto halfduplex, so no second parameter with the DE/RE pin is required!
ModbusClientRTU MB(Serial2);

// Define an onData handler function to receive the regular responses
// Arguments are Modbus server ID, the function code requested, the message data and length of it, 
// plus a user-supplied token to identify the causing request
void handleData(ModbusMessage response, uint32_t token) 
{
  LOG_N("Response: serverID=%d, FC=%d, Token=%08X, length=%d:\n", response.getServerID(), response.getFunctionCode(), token, response.size());
  HEXDUMP_N("Data", response.data(), response.size());
  //Serial.println(response.data());
  uint16_t val;
  /*for (int i=3;i<response.size();i+=2){
    response.get(i,val);
    //Serial.println(val);
  }*/
}

// Define an onError handler function to receive error responses
// Arguments are the error code returned and a user-supplied token to identify the causing request
void handleError(Error error, uint32_t token) 
{
  // ModbusError wraps the error code and provides a readable error message for it
  ModbusError me(error);
  LOG_E("Error response: %02X - %s\n", (int)me, (const char *)me);
}
uint32_t Token = 1111;
// Setup() - initialization happens here
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Set up Serial2 connected to Modbus RTU
// (Fill in your data here!)
  //Serial2.begin(19200, SERIAL_8N1, GPIO_NUM_17, GPIO_NUM_16);
  Serial2.begin(9600,SERIAL_8N1,16,17);
// Set up ModbusRTU client.
// - provide onData handler function
  MB.onDataHandler(&handleData);
// - provide onError handler function
  MB.onErrorHandler(&handleError);
// Set message timeout to 2000ms
  MB.setTimeout(2000);
// Start ModbusRTU background task
  MB.begin();

// We will first set the register to a known state, read the register, 
// then write to it and finally read it again to verify the change

// Set defined conditions first - write 0x1234 to the register
  
/*
  Error err = MB.addRequest(Token++, 1, WRITE_HOLD_REGISTER, 10, 0x1234);
  if (err!=SUCCESS) {
    ModbusError e(err);
    LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }*/

// Create request by
// (Fill in your data here!)
// - token to match the response with the request.
// - server ID = 1
// - function code = 0x03 (read holding register)
// - address to read = word 10
// - data words to read = 1
//
// If something is missing or wrong with the call parameters, we will immediately get an error code 
// and the request will not be issued
/*
// Read register
Error err = MB.addRequest(Token++, 1, READ_HOLD_REGISTER, 0x3000, 0x300E - 0x3000+1);
  if (err!=SUCCESS) {
    ModbusError e(err);
    LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }*/
/*
// Now write 0xBEEF to it
  err = MB.addRequest(Token++, 1, WRITE_HOLD_REGISTER, 10, 0xBEEF);
  if (err!=SUCCESS) {
    ModbusError e(err);
    LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }*/
/*
// Read it again to verify
  err = MB.addRequest(Token++, 1, READ_HOLD_REGISTER, 10, 1);
  if (err!=SUCCESS) {
    ModbusError e(err);
    LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }*/
/*
// Provoke an error just to show it
  err = MB.addRequest(Token++, 1, USER_DEFINED_44, 10, 1);
  if (err!=SUCCESS) {
    ModbusError e(err);
    LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }*/

// The output on the Serial Monitor will be (depending on your Modbus the data will be different):
//      __ OK __
//      [N] 156| main.cpp             [  27] handleData: Response: serverID=1, FC=6, Token=00000457, length=6:
//     [N] Data: @3FFB9878/6:
//        | 0000: 01 06 00 0A 12 34                                 |.....4          |
//      [N] 181| main.cpp             [  27] handleData: Response: serverID=1, FC=3, Token=00000458, length=5:
//      [N] Data: @3FFB9800/5:
//        | 0000: 01 03 02 12 34                                    |....4           |
//      [N] 205| main.cpp             [  27] handleData: Response: serverID=1, FC=6, Token=00000459, length=6:
//      [N] Data: @3FFB9800/6:
//        | 0000: 01 06 00 0A BE EF                                 |......          |
//      [N] 231| main.cpp             [  27] handleData: Response: serverID=1, FC=3, Token=0000045A, length=5:
//      [N] Data: @3FFB9800/5:
//        | 0000: 01 03 02 BE EF                                    |.....           |
//      [E] 255| main.cpp             [  37] handleError: Error response: 01 - Illegal function code
}

// loop() - nothing done here today!
void loop() {
  Error err = MB.addRequest(Token++, 1, READ_HOLD_REGISTER, 0x4000, 60);
  if (err!=SUCCESS) {
    ModbusError e(err);
    LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }
  delay(2000);
}
```

**Reply by Miq1** [reply]
Well, the code seems to be straightforward - removing all the commented-out lines would make it easier to read, though. . . 

I am not at the computer any more, so all I can offer today is some hints to narrow down the problem. 

In ``ModbusClientRTU.cpp`` around line 160 there is the check of the CRC for the received message. It should be helpful if you would add a line dumping out the received data there (``HEXDUMP_N("Received", msg.data(),msg.size());``).Then we can start analyzing what the response was, if it is completeand if the CRC in fact was correct.

**Reply by TLS1000** [reply]
Hi Michael, thanks to your library I succeeded in my goal for using the modbus in async way. About every two seconds a few hundred registers are read, while answering every single seconds to CAN-communication, and also handle a lot SPI traffic in different tasks.

I have split the 32 bit Token from the message into a generic 16-bit incrementing message counter, and the other 16 bit to tell the handler function in which array and index to store the message data. 

The problem with CRC-errors has been worked-around by chopping the requests into 'maximum 25 word' requests. I had to split large requests anyway into 128-byte-requests because of uart size, so I have limited it further to 50bytes per request.

I guess other users will experience the same issue when trying to request more than 25 words @9600bps. I'm too limited as a programmer to find the problem in libraries or in the arduino core for ESP32, but if there's an update just let me know and i'll test it. I'll be working on current setup for the next few weeks.

AND....THANK YOU!

**Reply by Miq1** [reply]
Good to read you found a way that is working for you. I would have wished to find the reason for your CRC issue, since we never encountered something similar before - except for the ``esp32-hal-uart.c`` stuff. There may be a bug hidden in the depths of the code. Well, if it is, someone else might encounter it again some time.

**Reply by Miq1** [reply]
Addendum: I tried myself once again with 100 requests of 125 registers each in the quickest possible succession - no CRC error at all. So I assume it might be a local issue on your bus - another server or even another client interfering or such. Sorry.

**Reply by TLS1000** [reply]
ok, thank you for verifying. Maybe it's indeed the slave device that's causing the CRC from time to time.

**Reply by TLS1000** [reply]
I see now that the CRC error is always accompagnied by timeout error. Could it be that your library sends the new request too fast after having received an answer? One loop of modbus tasks is about 10 requests. I also noticed that errors are grouped.

**Reply by Miq1** [reply]
On RTU Modbus, the interval between messages is determined by the bus speed. The bus silence time shall be 3.5x a character's transmission time. Such is the implementation:
```
// silent interval is at least 3.5x character time
    MR_interval = 35000000UL / MR_serial.baudRate();  // 3.5 * 10 bits * 1000 µs * 1000 ms / baud
```
This interval is maintained throughout the library. If conforming to the Modbus standard your server should be able to react on another request once the interval after the last response has passed.

You could try to use a larger value than ``35000000UL`` to make the interval longer to see if your server will do better. This is possible in a client only, though, a RTU server must obey the bus timing the (only) client does maintain.

**Reply by TLS1000** [reply]
I have no more timeout errors when I added a 10ms delay at the en of the handler function, and at the end of the error handler function. Now I only have CRC error every few seconds. I believe the host needs more time between calls, or maybe it's my driver ic (with auto-enable detection) https://imgaz.staticbg.com/thumb/large/oaupload/ser1/banggood/images/17/9E/29cbe5b1-42e2-4044-9d05-9c3e4fd95ff8.JPG.webp

where can I set this interval?

**Reply by Miq1** [reply]
> 
> 
> I see now that the CRC error is always accompagnied by timeout error. Could it be that your library sends the new request too fast after having received an answer? One loop of modbus tasks is about 10 requests. I also noticed that errors are grouped.

This massively hints to fragmented messages (the notorious ``esp32-hal-uart.c`` issue)! By default in the core code the first message is copied into the program buffer if at least 112 bytes are received, then the next 112 are awaited (or some relatively long timeout has passed). If now the bus timing suggest an end-of-message (more than the interval of bus silence is observed), the message as received is processed and will lead to a CRC error (since it is incomplete). Now the next request is sent and the result is garbage (the remainder of the previous response plus some) that is skipped. The next request so never gets its response - hence a Timeout error.

**Reply by Miq1** [reply]
> 
> 
> I have no more timeout errors when I added a 10ms delay at the en of the handler function, and at the end of the error handler function. Now I only have CRC error every few seconds. I believe the host needs more time between calls, or maybe it's my driver ic (with auto-enable detection) https://imgaz.staticbg.com/thumb/large/oaupload/ser1/banggood/images/17/9E/29cbe5b1-42e2-4044-9d05-9c3e4fd95ff8.JPG.webp
> 
These are exactly the adapters I normally will use - no issue known yet, as long as you keep your speed below ~300.000 baud.

If you revert your change with the RXFIFO_... definition, the library will fall back to the massive 16ms timeout after each transmission - more than enough for even the slowest server. But I seem to remember that you had timeouts with that in place as well...

**Reply by Miq1** [reply]
Do you have a logic analyzer or scope at hand? Then you could fetch the traffic back and forth on the UART side of the RTU adapter and see the timing.

**Reply by TLS1000** [reply]
I found the cause: in my arduino IDE boards manager I had the 1.0.4 version of the ESP32. I updated to 1.0.5 and I have no more errors, even at 64 words per request. I did the alteration in the options file and the patch in the hal-uart as suggested.

I'm sorry to have taken so much of your time with this stupid issue, I was under the assumption that the arduino IDE would ask me to do the board definition update when available.... If I could, I would send you some flowers :-)

**Comment by Miq1** [comment]
Congrats you found the cause! No worries about us looking into your issue - if it weren't fun, we would not do it. 😉

**Reply by Miq1** [reply]
@TLS1000 If you like, download a fresh copy of the library. I found another issue in ``esp32-hal-uart.c`` that can explain what you have experienced. In the latest version that is fixed, plus you need not to patch the core any more.

**Reply by TLS1000** [reply]
Hi Michael, thanks for the notice. I'll do the update and post my findings on the working. I'm always reluctant to make alterations to the arduino core so glad to hear you found another way.

**Reply by TLS1000** [reply]
Hi Michael, I advanced a bit in my project, but somehow I doubt if my writing of registers is correct. 

I would write 6 successive words via 0x10 function code to host adr 1, starting at the 0x0x4057 register:

```
uint16_t wData[]={3,5,7,9,3,2,9};
MB.addRequest(WRtoken++,1, WRITE_MULT_REGISTERS,0x4057,7,14, wData);
```
could you confirm this is correct? thank you

**Reply by Miq1** [reply]
Well, no... If you want to write an array of ``uint16_t`` words, you need to use the ``addRequest()`` variety #5 as is described [here](https://emodbus.github.io/modbusclient-common-api#setting-up-requests).
You wrote "I would write 6 successive words", but I reckon you meant 7, as your ``wData`` array has 7 entries and in your (wrong) ``addRequest()`` call you are giving 7 (or 14) as well. 😉

**Reply by TLS1000** [reply]
yes indeed, 7 words...
I'm confused by this example:
```
// Create request for
// (Fill in your data here!)
// - server ID = 1
// - function code = 0x16 (write multiple registers)
// - address to write = word 33ff
// - data words to write = see below
// - data bytes to write = see below
// - token to match the response with the request.
//
  uint16_t wData[] = { 0x1111, 0x2222, 0x3333, 0x4444, 0x5555, 0x6666 };

  err = MB.addRequest(Token++, 1, WRITE_MULT_REGISTERS, 33, 6, 12, wData);
  if (err!=SUCCESS) {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }
```

I suppose the function code for 'WRITE_MULT_REGISTERS' is not 0X16, but 0x10? But I don't see the problem with my statement, am I not doing exactly the same thing?

**Reply by Miq1** [reply]
Ough! You got me! The Modbus standard indeed is requiring both, a register count and a byte count:
![grafik](https://user-images.githubusercontent.com/25884124/113697388-6adbfd00-96d3-11eb-95de-18fbfc4530a5.png)
This is how the example was written. So my advice was wrong, the call you wrote is completely correct.

Sorry for the confusion I may have caused. 😊

**Reply by TLS1000** [reply]
well how about that! thanks for the confirmation.

**Reply by TLS1000** [reply]
Hi Michael, I continued working with your library using the 0X05, 0X06 and 0X10 function codes; great stuff! BUT when I disconnect the host it causes a reboot of the controller. This could endanger other running tasks. I set the timeout to 100ms but no use.

```
Error creating request: E8 - Request queue full
Error creating request: E7 - Parameter out of bounds
Error creating request: E8 - Request queue full
Error creating request: E8 - Request queue full
Error creating request: E8 - Request queue full
abort() was called at PC 0x400eb673 on core 0

ELF file SHA256: 0000000000000000

Backtrace: 0x40084f04:0x3ffb9450 0x40085179:0x3ffb9470 0x400eb673:0x3ffb9490 0x400eb6ba:0x3ffb94b0 0x400eada7:0x3ffb94d0 0x400eb0da:0x3ffb94f0 0x400d51e4:0x3ffb9510 0x400d569a:0x3ffb9530 0x400d56cd:0x3ffb9550 0x400d4aea:0x3ffb9570 0x40086179:0x3ffb95d0

Rebooting...
ets Jun  8 2016 00:22:57
```

**Reply by Miq1** [reply]
Interesting. The last few logged messages are showing that your program is continuing firing requests, despite you will have received an E3 connection error earlier.
The library should not abort, though, so it would be interesting to learn where in code the abort happens. Can you install the ESP32 exception decoder and list the stack trace it will generate in case of an abort?

**Reply by TLS1000** [reply]
sorry my mistake, I had an out of bounds reference to an array in my program. When I disconnect the host there are timeout-errors and that's it; the program keeps running fine.

**Reply by TLS1000** [reply]
Hi Michael, it's me again with another problem :-) I can switch on my device by sending 'err=MB.addRequest(WRtoken++,1, 0x05 ,0x2100 ,0xff00 );  //start' . But when I try to switch it off by sending 'err=MB.addRequest(WRtoken++,1, 0x05 ,0x2100 ,0x00ff );  //stop'. It gives 'error creating request, parameter out of bounds'. Do you see a problem with my second request? This error returns from the library and not from the host, am I correct? thankx

**Reply by Miq1** [reply]
You are right, it is the library complaining. As far as I know the OFF value for function code 0x05 is supposed to be 0x0000, not 0x00ff, as you used it.

**Reply by TLS1000** [reply]
the protocol description from the manufacturer specifies 0x00FF for switching off, and 0XFF00 for switching it on.... is there a workaround?

**Reply by TLS1000** [reply]
I tried 0x0000 without success...

**Reply by Miq1** [reply]
To be sure I checked the Modbus specs again: the manufacturer is wrong.

There is a workaround, but it is not an elegant one:
```
ModbusMessage msg;
msg.add((uint8_t) 1, WRITE_SINGLE_COIL, 0x2100, 0x00ff);
err = MB.addRequest(msg, WRtoken++);
```
I. E., you have to set up the message all yourself.

**Reply by TLS1000** [reply]
Hi Michael, I replaced 'WRITE_SINGLE_COIL' by 0x05 and on the serial monitor it returns '[1;33m[E] 19586| modbus.ino           [ 308] handleError: Error response: E3 - Function code mismatch' This is a response from the host I assume.

when I replace 'WRITE_SINGLE_COIL' (undefined) by 'WRITE_COIL' it returns [1;33m[E] 6606| modbus.ino           [ 308] handleError: Error response: E0 - Timeout

I verified: Serial.print("write_coil=");Serial.println(WRITE_COIL,HEX);    It returns 5! How can I have two different results?

**Reply by Miq1** [reply]
Argh! Of course it is WRITE_COIL! 0x05 it is as well, so the server seems to respond somewhat different. The E3 error signals the function code in the response was different to that of the request. 

Can you please use the ``onResponseHandler`` callback instead and print out the server's response by ``HEXDUMP_N("Response", response.data(), response.size());``? Then we will see what the server is responding in detail.

**Reply by Miq1** [reply]
Correction: we won't... The error message is generated internally and will hide the real response. 

The only way I see now is to recompile everything with ``#define LOG_LEVEL LOG_LEVEL_VERBOSE`` in ``Logging.h`` and dig through the tons of debug output. Alternatively, should you have a logic analyzer available, hook it to the UART pins where the RS485 adaptor is connected and look there for the message.
